Java 11 upgrade

carlosmunoz opened the pull request.

Commit:
add explicit import for class which clashes with a new JDK class

Commit:
upgrade the java version on the github maven action

Commit:
modify maven build settings to compile java 11 sources

This revision was also tested against a java 11 version of a GraalVM

xdavidson reviewed (commented):
8 days ago, Red Hat released openjdk11 on the ubi8 base image!
https://catalog.redhat.com/software/containers/ubi8/openjdk-11/5dd6a4b45a13461646f677f4

carlosmunoz commented:
@xdavidson this is good news. If we want to move forward with this upgrade, we will need to update our deployment images.

aprajshekhar commented:
@carlosmunoz I tried building and deploying this branch locally with OpenJDK 11. Build succeeds. However, I am getting the following exception when Karaf starts up -

```
org.ops4j.pax.logging.pax-logging-api [log4j2] ERROR : An exception occurred processing Appender PaxOsgi Ignored FQCN: org.apache.logging.log4j.spi.AbstractLogger
java.lang.NullPointerException
        at org.apache.logging.log4j.util.StackLocator9.calcLocation(StackLocator9.java:80)
        at org.apache.logging.log4j.util.StackLocatorUtil.calcLocation(StackLocatorUtil.java:94)
        at org.apache.logging.log4j.core.impl.Log4jLogEvent.getSource(Log4jLogEvent.java:693)
        at org.ops4j.pax.logging.log4j2.internal.spi.PaxLoggingEventImpl.getLocationInformation(PaxLoggingEventImpl.java:47)
        at org.ops4j.pax.logging.spi.support.PaxAppenderProxy.doAppend(PaxAppenderProxy.java:67)
        at org.ops4j.pax.logging.log4j2.internal.bridges.PaxOsgiAppender.append(PaxOsgiAppender.java:100)
        at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
        at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
        at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:543)
        at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:502)
        at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:485)
        at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:412)
        at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:63)
        at org.apache.logging.log4j.core.Logger.logMessage(Logger.java:153)
        at org.ops4j.pax.logging.log4j2.internal.PaxLoggerImpl.doLog0(PaxLoggerImpl.java:356)
        at org.ops4j.pax.logging.log4j2.internal.PaxLoggerImpl.doLog(PaxLoggerImpl.java:338)
        at org.ops4j.pax.logging.log4j2.internal.PaxLoggerImpl.error(PaxLoggerImpl.java:164)
        at org.ops4j.pax.logging.internal.TrackingLogger.error(TrackingLogger.java:130)
        at org.ops4j.pax.logging.spi.support.FrameworkHandler.doLog(FrameworkHandler.java:187)
        at org.ops4j.pax.logging.spi.support.FrameworkHandler.frameworkEvent(FrameworkHandler.java:149)
        at org.apache.felix.framework.EventDispatcher.invokeFrameworkListenerCallback(EventDispatcher.java:881)
        at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:830)
        at org.apache.felix.framework.EventDispatcher.run(EventDispatcher.java:1147)
        at org.apache.felix.framework.EventDispatcher.access$000(EventDispatcher.java:54)
        at org.apache.felix.framework.EventDispatcher$1.run(EventDispatcher.java:102)
        at java.base/java.lang.Thread.run(Thread.java:834)

```
Is it happening at your end as well?

aprajshekhar commented:
> 8 days ago, Red Hat released openjdk11 on the ubi8 base image!
> https://catalog.redhat.com/software/containers/ubi8/openjdk-11/5dd6a4b45a13461646f677f4

This is great. I was a bit concerned about JDK-11 not available for ubi8.

aprajshekhar reviewed (commented):
Apart from the exception mentioned earlier, all the functionality seems to be working as expected based on the smoke testing I did.

carlosmunoz commented:
> @carlosmunoz I tried building and deploying this branch locally with OpenJDK 11. Build succeeds. However, I am getting the following exception when Karaf starts up -
> 
> ```
> org.ops4j.pax.logging.pax-logging-api [log4j2] ERROR : An exception occurred processing Appender PaxOsgi Ignored FQCN: org.apache.logging.log4j.spi.AbstractLogger
> java.lang.NullPointerException
>         at org.apache.logging.log4j.util.StackLocator9.calcLocation(StackLocator9.java:80)
>         at org.apache.logging.log4j.util.StackLocatorUtil.calcLocation(StackLocatorUtil.java:94)
>         at org.apache.logging.log4j.core.impl.Log4jLogEvent.getSource(Log4jLogEvent.java:693)
>         at org.ops4j.pax.logging.log4j2.internal.spi.PaxLoggingEventImpl.getLocationInformation(PaxLoggingEventImpl.java:47)
>         at org.ops4j.pax.logging.spi.support.PaxAppenderProxy.doAppend(PaxAppenderProxy.java:67)
>         at org.ops4j.pax.logging.log4j2.internal.bridges.PaxOsgiAppender.append(PaxOsgiAppender.java:100)
>         at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
>         at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
>         at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
>         at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
>         at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:543)
>         at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:502)
>         at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:485)
>         at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:412)
>         at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:63)
>         at org.apache.logging.log4j.core.Logger.logMessage(Logger.java:153)
>         at org.ops4j.pax.logging.log4j2.internal.PaxLoggerImpl.doLog0(PaxLoggerImpl.java:356)
>         at org.ops4j.pax.logging.log4j2.internal.PaxLoggerImpl.doLog(PaxLoggerImpl.java:338)
>         at org.ops4j.pax.logging.log4j2.internal.PaxLoggerImpl.error(PaxLoggerImpl.java:164)
>         at org.ops4j.pax.logging.internal.TrackingLogger.error(TrackingLogger.java:130)
>         at org.ops4j.pax.logging.spi.support.FrameworkHandler.doLog(FrameworkHandler.java:187)
>         at org.ops4j.pax.logging.spi.support.FrameworkHandler.frameworkEvent(FrameworkHandler.java:149)
>         at org.apache.felix.framework.EventDispatcher.invokeFrameworkListenerCallback(EventDispatcher.java:881)
>         at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:830)
>         at org.apache.felix.framework.EventDispatcher.run(EventDispatcher.java:1147)
>         at org.apache.felix.framework.EventDispatcher.access$000(EventDispatcher.java:54)
>         at org.apache.felix.framework.EventDispatcher$1.run(EventDispatcher.java:102)
>         at java.base/java.lang.Thread.run(Thread.java:834)
> ```
> 
> Is it happening at your end as well?

@aprajshekhar I am not seeing this error

aprajshekhar reviewed (dismissed):
As I have not seen the error again and the smoke test was successful, let's go ahead with the changes.

carlosmunoz commented:
> As I have not seen the error again and the smoke test was successful, let's go ahead with the changes.

This will need a coordinated release with the pipelines, and changes in local dev environments. I will merge when/if given the go-ahead. (We also need to change Github's validations)

aprajshekhar commented:
> > As I have not seen the error again and the smoke test was successful, let's go ahead with the changes.
> 
> This will need a coordinated release with the pipelines, and changes in local dev environments. I will merge when/if given the go-ahead. (We also need to change Github's validations)

I agree. What I meant to say was this PR would not need Java 11 related changes except github actions :+1:

Commit:
rebase against master

The Module class needs to be explicitly imported since there is now a
Module class as part of the java.lang package.

Commit:
Merge branch 'master' into java_11_upgrade
